### ISSUE TITLE
Add Optuna-based hyperparameter tuning for model recipes

### ISSUE THREAD

**Issue by haukesteffen** [haukesteffen/TabularShenanigans#38]
## Summary
Add Optuna-backed hyperparameter tuning so the repository can search stronger parameters for a selected model recipe and then train the best found candidate into the normal artifact layout.

## Why
The current presets are fixed baselines. That is fine for bootstrapping the workflow, but it leaves clear performance on the table once a model family proves promising.

On competitions like `playground-series-s6e3`, the current results suggest model family choice matters, and tuning will become more valuable once stronger families such as CatBoost and LightGBM exist. We should introduce tuning as a thin layer over the existing CV and artifact system rather than as ad hoc scripts.

## Scope
- Integrate Optuna as an optional local hyperparameter tuner.
- Support tuning one explicit `model_id` per study in the first iteration.
- Reuse the existing task-aware CV splitter, metric scoring, positive-label handling, and preprocessing resolution.
- Define model-specific search spaces in code, not in `config.yaml`.
- Write tuning artifacts under `artifacts/<competition_slug>/tune/<study_id>/`.
- Persist at least:
  - `study_manifest.json`
  - `study_summary.csv`
  - `trials.csv`
  - `best_params.json`
- Allow the best trial params to feed back into normal training so the tuned candidate produces the standard train artifacts under `artifacts/<competition_slug>/train/<run_id>/`.
- Document the tuning workflow and config contract in `README.md` and `docs/TECHNICAL_GUIDE.md`.

## Design Constraints
- Keep `config.yaml` as the only runtime config source.
- Do not introduce an arbitrary YAML search-space DSL.
- Search spaces should live in code next to model definitions so they stay versioned and reviewable.
- Derive optimization direction from `primary_metric`; do not make users configure `maximize` vs `minimize`.
- Keep the implementation thin over existing module functions instead of forking the train pipeline.
- Preserve fail-fast behavior.
- Preserve reproducibility with explicit Optuna seed/random state.
- Initial implementation should be local CPU only.
- Do not make stage-specific CLI work a prerequisite for the core tuning logic.

## Recommended Config Shape
If we add tuning now, prefer a small explicit block such as:

```yaml
competition_slug: playground-series-s6e3
task_type: binary
primary_metric: roc_auc

tuning:
  enabled: true
  model_id: ordinal_hgb
  n_trials: 50
  timeout_seconds: 3600
  random_state: 42
```

Notes:
- `model_id` should be required for tuning in the first iteration.
- Search spaces should remain code-owned.
- The tuned best candidate should still flow through the normal train artifact contract.

## Initial Supported Models
Start with model recipes where tuning is likely to matter and integration is straightforward:
- `onehot_logreg`
- `ordinal_hgb`
- `ordinal_randomforest`
- `ordinal_extratrees`

The design should make it easy to add `lightgbm` and `catboost` later without changing the artifact contract.

## Acceptance Criteria
- A user can enable Optuna tuning from `config.yaml` for one explicit `model_id`.
- The tuner evaluates trials using the repository’s existing CV logic and configured primary metric.
- The run writes persistent tuning artifacts with per-trial parameters, scores, and trial state.
- The best trial parameters can be retrained into the normal training artifact layout without hand-editing code.
- Unsupported or untunable model recipes fail with a clear error.
- `README.md` and `docs/TECHNICAL_GUIDE.md` describe the tuning workflow, config keys, and artifacts.
- Manual verification is run on `playground-series-s6e3` with `task_type: binary` and `primary_metric: roc_auc`.
- Manual verification is run on `playground-series-s5e10` with `task_type: regression` and `primary_metric: mse`.

## Out Of Scope
- Arbitrary search-space definitions in YAML
- Distributed tuning or Ray integration
- Blend-weight optimization
- Feature selection or feature engineering search
- Automated leaderboard-driven tuning decisions
- Multi-objective optimization
- Automatic submission of every tuned candidate

## Notes
- This issue is probably lower priority than issue #4 if the immediate goal is leaderboard movement, because external boosters may matter more than tuning the current sklearn baselines.
- This issue should coordinate with issue #15 so tuned candidates and their later leaderboard outcomes remain traceable.
- If issue #16 lands first, it should add a `tune` stage over this machinery rather than introducing a separate tuning system.
